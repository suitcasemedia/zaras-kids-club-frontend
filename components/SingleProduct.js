import { useQuery } from '@apollo/client';
import gql from 'graphql-tag';
import Head from 'next/head';
import styled from 'styled-components';
import DisplayError from './ErrorMessage';

const ProductStyles = styled.div`
  color: white;
  background-color: var(--purple);
  display: grid;
  grid-auto-columns: 1fr;
  grid-auto-flow: column;
  gap: 2rem;
  max-width: var(--maxWidth);
  img {
    width: 100%;
    height: 100%;
    object-fit: contain;
  }
`;
const SINGLE_ITEM_QUERY = gql`
  query SINGLE_ITEM_QUERY($id: ID!) {
    Product(where: { id: $id }) {
      name
      price
      description
      id
      photo {
        image {
          publicUrlTransformed
        }
      }
    }
  }
`;

export default function SingleProduct({ id }) {
  const { data, loading, error } = useQuery(SINGLE_ITEM_QUERY, {
    variables: {
      id,
    },
  });

  if (loading) return <p> Loading...</p>;
  if (error) return <DisplayError />;
  const { Product } = data;
 
  return (
    <ProductStyles>
      <Head>
        <title>Zara's {Product.name}</title>
      </Head>
      <img
        alt={Product.name}
        src={Product?.photo?.image?.publicUrlTransformed}
      />
      <div className="details">
        <h2>{Product.name}</h2>
        <p>{Product.description}</p>
      </div>
    </ProductStyles>
  );
}
