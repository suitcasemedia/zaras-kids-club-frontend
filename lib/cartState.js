import { createContext, useContext, useState } from 'react';

const LocalStateContext = createContext();
const LocalStateProvider = LocalStateContext.Provider;
function CartStateProvider({ children }) {
  // this is its own custom provider

  const [cartOpen, setCartOpen] = useState(false);
  function toggleCart() {
    setCartOpen(!cartOpen);
  }

  function closeCart() {
    setCartOpen(false);
  }

  function openCart() {
    setCartOpen(true);
  }

  return (
    <LocalStateProvider
      value={{ cartOpen, setCartOpen, toggleCart, closeCart, openCart }}
    >
      {children}
    </LocalStateProvider>
  );
}

// make a custom hook for acessing the cart local state

function useCart() {
  const all = useContext(LocalStateContext);
  return all;
}
export { CartStateProvider, useCart };
