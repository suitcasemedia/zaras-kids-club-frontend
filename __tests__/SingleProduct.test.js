import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import SingleProduct, { SINGLE_ITEM_QUERY } from '../components/SingleProduct';
import { fakeItem } from '../lib/testUtils';

const product = fakeItem();
const mocks = [
  {
    // when the query is called with these variables
    request: {
      query: SINGLE_ITEM_QUERY,
      variables: { id: '123' },
    },

    // return this fake data (mocked data)
    result: {
      data: {
        Product: product,
      },
    },
  },
];

describe('<Single Product />', () => {
  it('renders with proper data', async () => {
    const { container, debug } = render(
      <MockedProvider mocks={mocks}>
        <SingleProduct id="123" />
      </MockedProvider>
    );

    await screen.findByTestId('single-product');

    expect(container).toMatchSnapshot();

    expect(container.querySelector('h2')).toHaveTextContent(product.name);
    expect(container.querySelector('p')).toHaveTextContent(product.description);
    expect(container.querySelector('img')).toHaveAttribute(
      'src',
      product.photo.image.publicUrlTransformed
    );
  });

  it('errors with a not found item', async () => {
    const notFoundMocks = [
      {
        request: {
          query: SINGLE_ITEM_QUERY,
          variables: { id: '123' },
        },
        result: {
          errors: [{ message: 'Items Not Found!' }],
        },
      },
    ];
    const { container, debug } = render(
      <MockedProvider mocks={notFoundMocks}>
        <SingleProduct id="123" />
      </MockedProvider>
    );

    await screen.findByTestId('graphql-error');

    expect(container).toHaveTextContent('Oh oh!');
    expect(container).toHaveTextContent('Items Not Found!');
    expect(container).toMatchSnapshot();
  });
  it('Errors out when an item is not found', async () => {
    const errorMock = [
      {
        request: {
          query: SINGLE_ITEM_QUERY,
          variables: { id: '123' },
        },
        result: {
          errors: [{ message: 'Items Not Found!' }],
        },
      },
    ];
    const { container, debug } = render(
      <MockedProvider mocks={errorMock}>
        <SingleProduct id="123" />
      </MockedProvider>
    );
    await screen.findByTestId('graphql-error');

    expect(container).toHaveTextContent('Items Not Found!');
    expect(container).toMatchSnapshot();
  });

  it('Renders loading state when the query is fired', async () => {
    const { container, debug } = render(
      <MockedProvider mocks={mocks}>
        <SingleProduct id="123" />
      </MockedProvider>
    );

    expect(container).toHaveTextContent('Loading...');
    expect(container).toMatchSnapshot();
  });
});
