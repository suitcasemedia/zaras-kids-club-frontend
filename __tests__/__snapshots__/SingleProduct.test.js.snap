// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Single Product /> Errors out when an item is not found 1`] = `
<div>
  <div
    class="ErrorMessage__ErrorStyles-sc-11u5fgj-0 fDAgpB"
    data-testid="graphql-error"
  >
    <p>
      <strong>
        Oh oh!
      </strong>
      Items Not Found!
    </p>
  </div>
</div>
`;

exports[`<Single Product /> Renders loading state when the query is fired 1`] = `
<div>
  <p>
     Loading...
  </p>
</div>
`;

exports[`<Single Product /> errors with a not found item 1`] = `
<div>
  <div
    class="ErrorMessage__ErrorStyles-sc-11u5fgj-0 fDAgpB"
    data-testid="graphql-error"
  >
    <p>
      <strong>
        Oh oh!
      </strong>
      Items Not Found!
    </p>
  </div>
</div>
`;

exports[`<Single Product /> renders with proper data 1`] = `
<div>
  <div
    class="SingleProduct__ProductStyles-sc-1kapz8w-0 cDdiNc"
    data-testid="single-product"
  >
    <img
      alt="dogs are best"
      src="dog.jpg"
    />
    <div
      class="details"
    >
      <h2>
        dogs are best
      </h2>
      <p>
        dogs
      </p>
    </div>
  </div>
</div>
`;
